ntbs-service/DataMigration/NotificationMapper.cs

```diff
@@ -271,23 +256,53 @@ private async Task<IEnumerable<IList<Notification>>> GetGroupedResultsAsNotifica
             treatmentEvents.AddRange(notificationOutcomeEvents);
 
             // If there is a death date but no death treatment event, add one in
-            if (rawNotification.DeathDate != null
-                && !treatmentEvents.Any(e => _diedOutcomeIds.Contains(e.TreatmentOutcomeId ?? 0)))
+            if (rawNotification.DeathDate != null && !treatmentEvents.Any(e => e.TreatmentEventIsDeathEvent))
             {
-                treatmentEvents.Add(CreateDerivedDeathEvent(rawNotification));
+                treatmentEvents.Add(await CreateDerivedDeathEvent(rawNotification));
             }
 
             return treatmentEvents;
         }
 
-        private TreatmentEvent CreateDerivedDeathEvent(MigrationDbNotification rawNotification) =>
-            new TreatmentEvent
+        private void AlterStillOnTreatmentEventDateIfEndingOutcomePresent(Notification notification)
+        {
+            var lastEndingEvent = notification.TreatmentEvents.Where(te => te.TreatmentEventIsEndingEvent)
+                .GetMostRecentTreatmentEvent();
+            if (!notification.ClinicalDetails.StartingDate.HasValue || lastEndingEvent == null || !lastEndingEvent.EventDate.HasValue)
             {
-                EventDate = rawNotification.DeathDate,
-                TreatmentEventType = TreatmentEventType.TreatmentOutcome,
-                TreatmentOutcomeId = _postMortemOutcomeType.TreatmentOutcomeId,
-                TreatmentOutcome = _postMortemOutcomeType
-            };
+                return;
+            }
+
+            var expectedEndingOutcomeYear = GetExpectedEndingOutcomeYear(notification, lastEndingEvent);
+            var stillOnTreatmentEvent = notification.TreatmentEvents.OrderForEpisodes().LastOrDefault(o => o.TreatmentOutcome?.TreatmentOutcomeSubType == TreatmentOutcomeSubType.StillOnTreatment);
+
+            if (TreatmentOutcomesHelper.GetTreatmentOutcomeAtXYears(notification, expectedEndingOutcomeYear)?.TreatmentOutcomeId != lastEndingEvent.TreatmentOutcomeId
```

Comment:
The logic here appears to be that we will update a still on treatment event if
a) There is an ending event before a still on treatment event
b) The outcome following the ending event doesn't match the ending event

However, I think we need a third condition:
c) There are no other events other than still on treatment events after the ending event.

Otherwise, we may see weird behaviour when treatment has genuinely restarted